Source/data/iterators.hpp

```diff
@@ -5,230 +5,251 @@
+	 */
+	[[nodiscard]] unsigned column() const
 	{
-		return {};
+		return column_;
 	}
 };
 
-class RecordsRange {
-	const char *const begin_;
+class RecordIterator {
+	GetFieldResult state_;
 	const char *const end_;
+	unsigned row_ = 0;
 
 public:
-	RecordsRange(std::string_view characters)
-	    : begin_(characters.data())
-	    , end_(characters.data() + characters.size())
+	using iterator_category = std::input_iterator_tag;
+	using value_type = RecordIterator;
+	using reference_type = value_type &;
+
+	RecordIterator()
+	    : state_(nullptr, GetFieldResult::Status::EndOfFile)
+	    , end_(nullptr)
 	{
 	}
 
-	class Iterator {
-		GetFieldResult state_;
-		const char *const end_;
-
-	public:
-		using iterator_category = std::forward_iterator_tag;
-		using value_type = FieldsInRecordRange;
-
-		Iterator()
-		    : state_(nullptr, GetFieldResult::Status::EndOfFile)
-		    , end_(nullptr)
-		{
-		}
-
-		Iterator(const char *begin, const char *end)
-		    : state_(begin)
-		    , end_(end)
-		{
-		}
+	RecordIterator(const char *begin, const char *end, bool skippedHeader)
+	    : state_(begin)
+	    , end_(end)
+	    , row_(skippedHeader ? 1 : 0)
+	{
+	}
 
-		[[nodiscard]] bool operator==(const Iterator &rhs) const
-		{
-			return state_.next == rhs.state_.next;
-		}
+	[[nodiscard]] bool operator==(const RecordIterator &rhs) const
+	{
+		return state_.next == rhs.state_.next;
+	}
 
-		[[nodiscard]] bool operator!=(const Iterator &rhs) const
-		{
-			return !(*this == rhs);
-		}
+	[[nodiscard]] bool operator!=(const RecordIterator &rhs) const
+	{
+		return !(*this == rhs);
+	}
 
-		Iterator &operator++()
-		{
-			return *this += 1;
-		}
+	RecordIterator &operator++()
+	{
+		return *this += 1;
+	}
 
-		Iterator &operator+=(unsigned increment)
-		{
-			if (!state_.endOfRecord())
-				state_ = DiscardRemainingFields(state_.next, end_);
-
-			if (state_.endOfFile()) {
-				state_.next = nullptr;
-			} else {
-				if (increment > 0)
-					state_ = DiscardMultipleRecords(state_.next, end_, increment - 1);
-				// We use Status::ReadyToRead as a marker in case the Field iterator is never
-				//  used, so the next call to operator+= will advance past the current record
-				state_.status = GetFieldResult::Status::ReadyToRead;
+	RecordIterator &operator+=(unsigned increment)
+	{
+		if (!state_.endOfRecord())
+			state_ = DiscardRemainingFields(state_.next, end_);
+
+		if (state_.endOfFile()) {
+			state_.next = nullptr;
+		} else {
+			if (increment > 0) {
+				row_ += increment;
+				state_ = DiscardMultipleRecords(state_.next, end_, increment - 1);
 			}
-			return *this;
+			// We use Status::ReadyToRead as a marker in case the Field iterator is never
+			//  used, so the next call to operator+= will advance past the current record
+			state_.status = GetFieldResult::Status::ReadyToRead;
 		}
+		return *this;
+	}
 
-		[[nodiscard]] value_type operator*()
-		{
-			return { &state_, end_ };
-		}
-	};
+	[[nodiscard]] reference_type operator*()

```

Comment:
This suggests that `RecordIterator` should not be an iterator.

Perhaps the following structure:

* `RecordsRange`: `begin()/end() RecordIterator`
* `RecordIterator`: `value_type = FieldsRange`, `row()` method
* `FieldsRange`: `begin() / end() FieldIterator`, also `row()` method here
* `FieldIterator`: `value_type = std::string_view`, `column()` method

Then the iteration would be as follows:

```c++
for (FieldsRange fields : dataFile.records()) {
  FieldIterator it = fields.begin();
  FieldIterator end = fields.end();
  
  // fields.row(),  it.column()
}
```

Reply:
I ended up going with a slightly different approach, the RecordsRange/RangeIterator/FieldsRange/FieldIterator structure was the thing I wasn't really happy with.

Now a DataFile is a range of Records (using a RecordIterator as a forward_iterator). A Record is a range of Fields (using a FieldIterator as a forward_iterator).

This lets a simple loop over a file be written as:
```cpp
DataFile file;
for (Record record : file) {
  for (Field field : record) { // alternatively `for (std::string_view field : record) {}` can be used if the value will only ever be used as a string and row/column numbers don't matter
    if (field == "string_view"sv)
      // can treat the value as if it was a string_view
    // can use field.parseInt(dest) or expected = field.asInt<int8_t>() to get a numeric value
    // field.row() and field.column() return the current file position
  }
}
```

Having the row/column available at the field level turned out to be really helpful for error messages
![2023-08-28T233236_devenv](https://github.com/diasurgical/devilutionX/assets/357684/b8cba70b-ca55-46b5-afa8-b1c5a6691812)

Reply:
Looks great, except for the implicit `string_view` conversion (left a comment) 🎉